power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp

```diff
@@ -6,16 +6,306 @@
+            serialize(true);
+        }
+        return {msgpack_buffer_.data(), msgpack_buffer_.size()};
+    }
+
+    std::string const& get_json(Idx indent) {
+        auto const json_document = nlohmann::json::from_msgpack(get_msgpack());
+        json_buffer_ = json_document.dump(static_cast<int>(indent));
+        return json_buffer_;
+    }
+
+  private:
+    MetaDataset const* dataset_;
+    bool is_batch_;
+    Idx batch_size_;
+    Idx n_components_;
+    std::vector<ScenarioBuffer> scenario_buffers_;   // list of scenarios, then list of components, omit empty
+    std::vector<ComponentBuffer> component_buffers_; // list of components, then all scenario flatten
+    // msgpack pakcer
+    msgpack::sbuffer msgpack_buffer_{};
+    msgpack::packer<msgpack::sbuffer> packer_;
+    bool use_compact_list_{};
+    std::map<MetaComponent const*, std::vector<MetaAttribute const*>> attributes_;
+    // json
+    std::string json_buffer_;
+
+    void store_buffers(char const** components, Idx const* elements_per_scenario, Idx const** indptrs,
+                       void const** data) {
+        scenario_buffers_.resize(batch_size_);
+        for (Idx scenario = 0; scenario != batch_size_; ++scenario) {
+            scenario_buffers_[scenario] =
+                create_scenario_buffer_view(components, elements_per_scenario, indptrs, data, scenario);
+        }
+        component_buffers_ =
+            create_scenario_buffer_view(components, elements_per_scenario, indptrs, data).component_buffers;
+    }
+
+    ScenarioBuffer create_scenario_buffer_view(char const** components, Idx const* elements_per_scenario,
+                                               Idx const** indptrs, void const** data, Idx scenario = -1) {
+        ScenarioBuffer scenario_buffer{};
+        Idx const begin_scenario = scenario < 0 ? 0 : scenario;
+        Idx const end_scenario = scenario < 0 ? batch_size_ : begin_scenario + 1;
+        for (Idx component = 0; component != n_components_; ++component) {
+            ComponentBuffer component_buffer{};
+            component_buffer.component = &dataset_->get_component(components[component]);
+            if (elements_per_scenario[component] < 0) {
+                component_buffer.data =
+                    component_buffer.component->advance_ptr(data[component], indptrs[component][begin_scenario]);
+                component_buffer.size = indptrs[component][end_scenario] - indptrs[component][begin_scenario];
+            } else {
+                component_buffer.data = component_buffer.component->advance_ptr(
+                    data[component], elements_per_scenario[component] * begin_scenario);
+                component_buffer.size = elements_per_scenario[component] * (end_scenario - begin_scenario);
+            }
+            // only store the view if it is non-empty
+            if (component_buffer.size > 0) {
+                scenario_buffer.component_buffers.push_back(component_buffer);
+            }
```

Comment:
theoretically, an empty scenario could exist as well, right? (although it wouldn't make sense to add more than one to an update dataset)

Reply:
Empty scenario will be correctly serialized. This is about empty component. An empty component should not exist in the dict of one scenario. If there are no components in a scenario, the scenario self should exist as an empty dict.